File: python/forge-server/tests/contract/corpus.py
"""A typed reading of ``contract/corpus.json``, and the rules that keep it
honest.

Nothing here knows about HTTP. A driver supplies the transport: it builds the
fixture, turns a :class:`Request` into whatever its transport sends, and hands
the response back to the matcher.

Reading is strict. An unknown field is an error, because a corpus that is
half-read is worse than one that fails to load: the unread half looks like
coverage.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
from typing import Any

# corpus.py -> contract -> tests -> forge-server -> python -> the repo root
CORPUS_PATH = Path(__file__).resolve().parents[4] / "contract" / "corpus.json"

#: An expectation the case did not author. ``null`` is itself a matcher, so
#: absence needs a value of its own.
ABSENT = object()

#: Transport id of the Python HTTP driver.
PYTHON_HTTP = "python-http"


class CorpusError(Exception):
    """A corpus that cannot be read, or cannot be run honestly."""


class Kind(str, Enum):
    """What a case exercises, which decides the shape of its steps."""

    HTTP = "http"
    SSE = "sse"
    WS = "ws"


class Auth(str, Enum):
    """How a request carries its identity."""

    NONE = "none"
    BEARER = "bearer"
    QUERY = "query"


@dataclass(frozen=True)
class FixtureUser:
    name: str
    #: Plaintext. How a driver stores it is its own business.
    password: str
    roles: list[str]


@dataclass(frozen=True)
class FixtureAuth:
    #: Auth on. Off means every endpoint is open and the identity is anonymous.
    enabled: bool
    users: list[FixtureUser]


@dataclass(frozen=True)
class FixtureComponents:
    #: Written to ``manifest.json`` in the components directory.
    manifest: Any
    #: Written beside it: filename to content.
    files: dict[str, str]


@dataclass(frozen=True)
class FixtureFrontend:
    #: Written to the static frontend directory: filename to content.
    files: dict[str, str]


@dataclass(frozen=True)
class Fixture:
    """The server state the cases assume. See ``contract/README.md`` for the
    behaviour each named action must have."""

    app: str
    auth: FixtureAuth
    docstore: bool
    events: bool
    actions: list[str]
    components: FixtureComponents
    frontend: FixtureFrontend


@dataclass(frozen=True)
class Request:
    """One request, as it goes on the wire."""

    method: str
    #: A raw URI path, sent verbatim — already percent-encoded where it needs
    #: to be.
    path: str
    #: Query parameters. The driver encodes the values.
    query: dict[str, str]
    #: Extra headers, on top of whatever ``auth`` adds.
    headers: dict[str, str]
    auth: Auth
    #: A JSON request body, sent with a JSON content type. ``None`` = no body.
    body: Any


@dataclass(frozen=True)
class Connect:
    """A websocket handshake."""

    path: str
    query: dict[str, str]
    auth: Auth


@dataclass(frozen=True)
class Expect:
    """What must come back."""

    #: The HTTP status. A transport without status lines maps it through the
    #: contract's error kinds — see ``contract/README.md``.
    status: int
    #: Header name (lower-case) to a matcher over its value.
    headers: dict[str, Any]
    #: Matcher over the parsed JSON body, or :data:`ABSENT`.
    body: Any
    #: Matcher over the raw body for a response that is not JSON, or
    #: :data:`ABSENT`.
    text: Any


@dataclass(frozen=True)
class RequestStep:
    """Send one request and check the response."""

    request: Request
    expect: Expect | None


@dataclass(frozen=True)
class ConnectStep:
    """Open a websocket. An ``expect`` means the handshake must be refused."""

    connect: Connect
    expect: Expect | None


@dataclass(frozen=True)
class SendStep:
    """Send a JSON frame on the open socket."""

    frame: Any


@dataclass(frozen=True)
class AwaitFrameStep:
    """The next frame on the socket must match."""

    matcher: Any


@dataclass(frozen=True)
class AwaitEventStep:
    """The next event on the stream must have this topic and match this data."""

    topic: str
    data: Any


Step = RequestStep | ConnectStep | SendStep | AwaitFrameStep | AwaitEventStep


@dataclass(frozen=True)
class Case:
    """One contract case."""

    id: str
    title: str
    kind: Kind
    #: Why the case is written the way it is. Not asserted on.
    note: str | None
    #: Transports that must run this case.
    applies: list[str]
    #: Transports that cannot serve it, and what stops them.
    inapplicable: dict[str, str]
    steps: list[Step] = field(default_factory=list)

    def applies_to(self, transport: str) -> bool:
        """Whether this transport must run the case."""
        return transport in self.applies


@dataclass(frozen=True)
class Corpus:
    """One authored contract corpus."""

    #: Contract version the cases describe (``docs/api-contract.md``).
    contract_version: str
    #: Every transport a case must account for.
    transports: list[str]
    #: Substitution table for ``${name}`` in paths, bodies and expectations.
    vars: dict[str, str]
    #: The server every driver must build before running a case.
    fixture: Fixture
    #: The contract cases, in authored order.
    cases: list[Case]

    @classmethod
    def load(cls, path: Path = CORPUS_PATH) -> "Corpus":
        """Parse and validate the authored corpus."""
        try:
            raw = path.read_text()
        except OSError as e:
            raise CorpusError(f"cannot read {path}: {e}") from e
        return cls.parse(raw)

    @classmethod
    def parse(cls, text: str) -> "Corpus":
        """Parse and validate a corpus from JSON."""
        try:
            raw = json.loads(text)
        except json.JSONDecodeError as e:
            raise CorpusError(f"corpus is not readable: {e}") from e
        corpus = _corpus(raw)
        corpus.validate()
        return corpus

    def cases_for(self, transport: str) -> list[Case]:
        """Cases a transport must run, in authored order."""
        return [case for case in self.cases if case.applies_to(transport)]

    def validate(self) -> None:
        """Reject a corpus that cannot be run honestly.

        The rule that matters: every case accounts for every transport, so a
        coverage gap has to be written down rather than left out.
        """
        if not self.transports:
            raise CorpusError("corpus declares no transports")
        seen: set[str] = set()
        for case in self.cases:
            if case.id in seen:
                raise CorpusError(f"duplicate case id {case.id!r}")
            seen.add(case.id)
            self._validate_applicability(case)
            self._validate_steps(case)

    def _validate_applicability(self, case: Case) -> None:
        for transport in case.applies:
            if transport not in self.transports:
                raise CorpusError(
                    f"case {case.id!r} applies to unknown transport {transport!r}"
                )
        for transport, reason in case.inapplicable.items():
            if transport not in self.transports:
                raise CorpusError(
                    f"case {case.id!r} excuses unknown transport {transport!r}"
                )
            if not reason.strip():
                raise CorpusError(f"case {case.id!r} excuses {transport!r} with no reason")
            if case.applies_to(transport):
                raise CorpusError(
                    f"case {case.id!r} both applies to and excuses {transport!r}"
                )
        for transport in self.transports:
            if not case.applies_to(transport) and transport not in case.inapplicable:
                raise CorpusError(
                    f"case {case.id!r} says nothing about transport {transport!r} — "
                    "list it under `applies` or give a reason under `inapplicable`"
                )

    def _validate_steps(self, case: Case) -> None:
        if not case.steps:
            raise CorpusError(f"case {case.id!r} has no steps")
        first = case.steps[0]
        if case.kind is Kind.HTTP:
            for step in case.steps:
                if not isinstance(step, RequestStep):
                    raise CorpusError(
                        f"case {case.id!r} is kind `http`, so every step must be a request"
                    )
        elif case.kind is Kind.SSE:
            if not isinstance(first, RequestStep):
                raise CorpusError(
                    f"case {case.id!r} is kind `sse`, so its first step must be the "
                    "request that opens the stream"
                )
            # The stream's own response has no body to read — it is the stream.
            # A driver would drop a body expectation authored here without a
            # word, which is the silent gap this corpus exists to stop. Assert
            # on the events instead, with `await_event`.
            authored = first.expect is not None and (
                first.expect.body is not ABSENT or first.expect.text is not ABSENT
            )
            if authored:
                raise CorpusError(
                    f"case {case.id!r} expects a body from the request that opens the "
                    "stream; only its status and headers can be checked"
                )
            for step in case.steps:
                if isinstance(step, (ConnectStep, SendStep)):
                    raise CorpusError(
                        f"case {case.id!r} is kind `sse`; a stream cannot be connected "
                        "to or sent on"
                    )
        else:
            if not isinstance(first, ConnectStep):
                raise CorpusError(
                    f"case {case.id!r} is kind `ws`, so its first step must be a connect"
                )
            for step in case.steps[1:]:
                if isinstance(step, (ConnectStep, AwaitEventStep)):
                    raise CorpusError(
                        f"case {case.id!r} connects once, and awaits frames rather "
                        "than events"
                    )


# -- reading ---------------------------------------------------------------
#
# One helper per shape, so an unreadable corpus says which field it tripped on.


def _fields(raw: Any, where: str, known: set[str]) -> dict:
    if not isinstance(raw, dict):
        raise CorpusError(f"{where}: expected an object, got {type(raw).__name__}")
    unknown = sorted(set(raw) - known)
    if unknown:
        raise CorpusError(f"{where}: unknown field(s) {', '.join(repr(k) for k in unknown)}")
    return raw


def _required(raw: dict, key: str, where: str, kind: type | tuple[type, ...]) -> Any:
    if key not in raw:
        raise CorpusError(f"{where}: missing {key!r}")
    return _typed(raw[key], f"{where}.{key}", kind)


def _optional(raw: dict, key: str, where: str, kind: type | tuple[type, ...], default: Any) -> Any:
    if key not in raw:
        return default
    return _typed(raw[key], f"{where}.{key}", kind)


def _typed(value: Any, where: str, kind: type | tuple[type, ...]) -> Any:
    if kind is not Any and not isinstance(value, kind):
        names = kind if isinstance(kind, tuple) else (kind,)
        raise CorpusError(
            f"{where}: expected {' or '.join(k.__name__ for k in names)}, "
            f"got {type(value).__name__}"
        )
    return value


def _str_map(raw: Any, where: str) -> dict[str, str]:
    mapping = _typed(raw, where, dict)
    for key, value in mapping.items():
        _typed(value, f"{where}.{key}", str)
    return dict(mapping)


def _str_list(raw: Any, where: str) -> list[str]:
    items = _typed(raw, where, list)
    return [_typed(item, f"{where}[{i}]", str) for i, item in enumerate(items)]


def _enum(raw: Any, where: str, kind: type[Enum]) -> Any:
    value = _typed(raw, where, str)
    try:
        return kind(value)
    except ValueError:
        allowed = ", ".join(repr(member.value) for member in kind)
        raise CorpusError(f"{where}: {value!r} is not one of {allowed}") from None


def _corpus(raw: Any) -> Corpus:
    obj = _fields(raw, "corpus", {"contract_version", "transports", "vars", "fixture", "cases"})
    cases = _required(obj, "cases", "corpus", list)
    return Corpus(
        contract_version=_required(obj, "contract_version", "corpus", str),
        transports=_str_list(_required(obj, "transports", "corpus", list), "corpus.transports"),
        vars=_str_map(_required(obj, "vars", "corpus", dict), "corpus.vars"),
        fixture=_fixture(_required(obj, "fixture", "corpus", dict)),
        cases=[_case(item, i) for i, item in enumerate(cases)],
    )


def _fixture(raw: Any) -> Fixture:
    where = "fixture"
    obj = _fields(
        raw, where, {"app", "auth", "docstore", "events", "actions", "components", "frontend"}
    )
    components = _fields(
        _required(obj, "components", where, dict), f"{where}.components", {"manifest", "files"}
    )
    frontend = _fields(
        _required(obj, "frontend", where, dict), f"{where}.frontend", {"files"}
    )
    return Fixture(
        app=_required(obj, "app", where, str),
        auth=_fixture_auth(_required(obj, "auth", where, dict)),
        docstore=_required(obj, "docstore", where, bool),
        events=_required(obj, "events", where, bool),
        actions=_str_list(_required(obj, "actions", where, list), f"{where}.actions"),
        components=FixtureComponents(
            manifest=_required(components, "manifest", f"{where}.components", Any),
            files=_str_map(
                _required(components, "files", f"{where}.components", dict),
                f"{where}.components.files",
            ),
        ),
        frontend=FixtureFrontend(
            files=_str_map(
                _required(frontend, "files", f"{where}.frontend", dict),
                f"{where}.frontend.files",
            )
        ),
    )


def _fixture_auth(raw: Any) -> FixtureAuth:
    where = "fixture.auth"
    obj = _fields(raw, where, {"enabled", "users"})
    users = _required(obj, "users", where, list)
    return FixtureAuth(
        enabled=_required(obj, "enabled", where, bool),
        users=[_fixture_user(user, i) for i, user in enumerate(users)],
    )


def _fixture_user(raw: Any, index: int) -> FixtureUser:
    where = f"fixture.auth.users[{index}]"
    obj = _fields(raw, where, {"name", "password", "roles"})
    return FixtureUser(
        name=_required(obj, "name", where, str),
        password=_required(obj, "password", where, str),
        roles=_str_list(_optional(obj, "roles", where, list, []), f"{where}.roles"),
    )


def _case(raw: Any, index: int) -> Case:
    where = f"cases[{index}]"
    obj = _fields(
        raw, where, {"id", "title", "kind", "note", "applies", "inapplicable", "steps"}
    )
    case_id = _required(obj, "id", where, str)
    steps = _required(obj, "steps", where, list)
    return Case(
        id=case_id,
        title=_required(obj, "title", where, str),
        kind=_enum(_optional(obj, "kind", where, str, "http"), f"{where}.kind", Kind),
        note=_optional(obj, "note", where, str, None),
        applies=_str_list(_required(obj, "applies", where, list), f"{where}.applies"),
        inapplicable=_str_map(
            _optional(obj, "inapplicable", where, dict, {}), f"{where}.inapplicable"
        ),
        steps=[_step(step, f"case {case_id!r} step {i}") for i, step in enumerate(steps)],
    )


def _step(raw: Any, where: str) -> Step:
    if not isinstance(raw, dict):
        raise CorpusError(f"{where}: expected an object, got {type(raw).__name__}")
    if "request" in raw:
        obj = _fields(raw, where, {"request", "expect"})
        return RequestStep(
            request=_request(_required(obj, "request", where, dict), where),
            expect=_expect_or_none(obj, where),
        )
    if "connect" in raw:
        obj = _fields(raw, where, {"connect", "expect"})
        return ConnectStep(
            connect=_connect(_required(obj, "connect", where, dict), where),
            expect=_expect_or_none(obj, where),
        )
    if "send" in raw:
        return SendStep(frame=_fields(raw, where, {"send"})["send"])
    if "await_frame" in raw:
        return AwaitFrameStep(matcher=_fields(raw, where, {"await_frame"})["await_frame"])
    if "await_event" in raw:
        obj = _fields(raw, where, {"await_event"})
        event = _fields(
            _required(obj, "await_event", where, dict), f"{where}.await_event", {"topic", "data"}
        )
        return AwaitEventStep(
            topic=_required(event, "topic", f"{where}.await_event", str),
            data=_required(event, "data", f"{where}.await_event", Any),
        )
    raise CorpusError(
        f"{where}: not a step — expected one of `request`, `connect`, `send`, "
        "`await_frame`, `await_event`"
    )


def _request(raw: Any, where: str) -> Request:
    where = f"{where}.request"
    obj = _fields(raw, where, {"method", "path", "query", "headers", "auth", "body"})
    if "body" in obj and obj["body"] is None:
        raise CorpusError(f"{where}: a request with no body omits `body`")
    return Request(
        method=_required(obj, "method", where, str),
        path=_required(obj, "path", where, str),
        query=_str_map(_optional(obj, "query", where, dict, {}), f"{where}.query"),
        headers=_str_map(_optional(obj, "headers", where, dict, {}), f"{where}.headers"),
        auth=_enum(_optional(obj, "auth", where, str, "none"), f"{where}.auth", Auth),
        body=obj.get("body"),
    )


def _connect(raw: Any, where: str) -> Connect:
    where = f"{where}.connect"
    obj = _fields(raw, where, {"path", "query", "auth"})
    return Connect(
        path=_required(obj, "path", where, str),
        query=_str_map(_optional(obj, "query", where, dict, {}), f"{where}.query"),
        auth=_enum(_optional(obj, "auth", where, str, "none"), f"{where}.auth", Auth),
    )


def _expect_or_none(raw: dict, where: str) -> Expect | None:
    if "expect" not in raw:
        return None
    where = f"{where}.expect"
    obj = _fields(raw["expect"], where, {"status", "headers", "body", "text"})
    return Expect(
        status=_required(obj, "status", where, int),
        headers=_typed(_optional(obj, "headers", where, dict, {}), f"{where}.headers", dict),
        body=obj.get("body", ABSENT),
        text=obj.get("text", ABSENT),
    )

File: crates/forge-tauri/tests/corpus.rs
//! The Rust IPC driver for the contract corpus (`contract/corpus.json`).
//!
//! It builds the fixture the corpus describes, then runs every case that
//! declares `rust-ipc` under `applies`. The case list lives in the corpus, not
//! here — this file only knows how to turn an authored request into an IPC
//! request and hand the response back to the matcher.
//!
//! There is no Tauri runtime in the loop: `Builder::try_state` assembles the
//! same plugin state the `setup` hook does, and `ForgeState::request` is the
//! same entry point the `plugin:forge|request` command calls. What a real app
//! adds on top is the invoke boundary, which carries the arguments verbatim.
//!
//! A case this transport cannot serve is declared inapplicable in the corpus,
//! with a reason. This driver has no skips: an authored expectation it cannot
//! check is a failure, because a silent pass is what the corpus exists to stop.

use std::collections::BTreeMap;

use forge_contract::{
    interpolate, interpolate_value, match_value, users_env, write_fixture_files, Auth, Case,
    Corpus, Expect, Fixture, Kind, Step, Vars, RUST_IPC,
};
use forge_tauri::{ActionCtx, AuthConfig, Builder, ForgeResponse, ForgeState};
use serde_json::{json, Value};

/// Driver-local: the corpus does not observe the signing secret.
const SECRET: &str = "0123456789abcdef0123456789abcdef";

#[tokio::test]
async fn corpus_rust_ipc() {
    let corpus = Corpus::load().expect("contract/corpus.json");

    // One state per fixture a case actually names, built the first time it is
    // needed: a fixture whose cases are all inapplicable here costs nothing.
    let mut harnesses: BTreeMap<&str, Harness> = BTreeMap::new();
    let mut failures = Vec::new();
    let mut ran = 0;
    for case in corpus.cases_for(RUST_IPC) {
        ran += 1;
        if !harnesses.contains_key(case.fixture.as_str()) {
            let harness = Harness::build(&corpus, &case.fixture).await;
            harnesses.insert(case.fixture.as_str(), harness);
        }
        let harness = &harnesses[case.fixture.as_str()];
        if let Err(why) = harness.run(case).await {
            failures.push(format!(
                "{} [{}]: {why}\n    ({})",
                case.id, case.fixture, case.title
            ));
        }
    }

    assert!(ran > 0, "no corpus case applies to {RUST_IPC}");
    assert!(
        failures.is_empty(),
        "{} of {ran} contract cases failed:\n  - {}",
        failures.len(),
        failures.join("\n  - ")
    );
}

/// The fixture the corpus describes, plus the token every case borrows.
struct Harness {
    state: ForgeState,
    vars: Vars,
    /// Kept alive: the doc store and the components directory live under it.
    _dir: tempfile::TempDir,
}

impl Harness {
    async fn build(corpus: &Corpus, name: &str) -> Self {
        let fixture = corpus
            .fixtures
            .get(name)
            .unwrap_or_else(|| panic!("the corpus has no fixture {name:?}"));
        let vars = corpus.vars();
        let dir = tempfile::tempdir().expect("tempdir");
        let data = dir.path().join("data");
        let components = dir.path().join("components");
        for path in [&data, &components] {
            std::fs::create_dir_all(path).expect("fixture dir");
        }
        // `fixture.frontend` has no counterpart here: the webview loads its
        // own assets, which is why the static cases are inapplicable.

        let mut builder = Builder::new(interpolate(&fixture.app, &vars).expect("app"));
        if let Some(fixture_components) = &fixture.components {
            // An authored manifest is written; an absent one is the point of
            // the fixture that leaves it out.
            if let Some(manifest) = &fixture_components.manifest {
                let manifest = interpolate_value(manifest, &vars).expect("manifest");
                std::fs::write(
                    components.join("manifest.json"),
                    serde_json::to_vec_pretty(&manifest).expect("manifest json"),
                )
                .expect("write manifest");
            }
            write_fixture_files(&components, &fixture_components.files, &vars).expect("components");
            builder = builder.with_components(&components);
        }
        if fixture.docstore {
            builder = builder.with_docstore(&data);
        }
        if fixture.auth.enabled {
            builder = builder.auth(auth_config(fixture, &vars));
        }
        for name in &fixture.actions {
            builder = register_action(builder, name);
        }
        // `fixture.events` needs no wiring: the bus is always live and the
        // plugin fans it out as Tauri events, which no case observes.
        let state = builder.try_state().expect("fixture state");

        let mut harness = Self {
            state,
            vars,
            _dir: dir,
        };
        let token = harness.login(fixture).await;
        harness.vars.insert("token".into(), token);
        harness
    }

    /// The one thing the driver does that the corpus does not describe: it
    /// needs a token before it can run a case that carries one.
    async fn login(&self, fixture: &Fixture) -> String {
        if !fixture.auth.enabled {
            return String::new();
        }
        let user = fixture.auth.users.first().expect("fixture user");
        let body = json!({
            "username": interpolate(&user.name, &self.vars).expect("user"),
            "password": interpolate(&user.password, &self.vars).expect("password"),
        });
        let res = self
            .state
            .request("POST", "/api/auth/login", Some(body), None)
            .await;
        assert_eq!(res.status, 200, "fixture login: {}", res.body);
        res.body
            .pointer("/data/token")
            .and_then(Value::as_str)
            .expect("login returns a token")
            .to_string()
    }

    async fn run(&self, case: &Case) -> Result<(), String> {
        if case.kind != Kind::Http {
            return Err(format!(
                "kind `{:?}` needs a stream this transport does not have; \
                 such a case belongs under `inapplicable`",
                case.kind
            ));
        }
        for (i, step) in case.steps.iter().enumerate() {
            let Step::Request { request, expect } = step else {
                return Err(format!("step {i}: kind `http` takes request steps only"));
            };
            let res = self.send(request, i).await?;
            self.check(expect.as_ref(), &res, i)?;
        }
        Ok(())
    }

    /// One authored request, as an IPC request. The path travels verbatim —
    /// the bridge percent-decodes each segment the way axum decodes a path
    /// parameter — and the token travels as an argument.
    async fn send(
        &self,
        request: &forge_contract::Request,
        step: usize,
    ) -> Result<ForgeResponse, String> {
        let at = |e: String| format!("step {step}: {e}");
        if !request.query.is_empty() {
            return Err(at("an IPC request carries no query string, so it cannot \
                           carry a query parameter; a case that needs one belongs \
                           under `inapplicable`"
                .into()));
        }
        let path = interpolate(&request.path, &self.vars).map_err(at)?;
        let body = match &request.body {
            Some(body) => Some(interpolate_value(body, &self.vars).map_err(at)?),
            None => None,
        };
        let token = self.token(request).map_err(at)?;
        Ok(self
            .state
            .request(&request.method, &path, body, token.as_deref())
            .await)
    }

    /// The token this request carries. `bearer` takes the fixture's; an
    /// authored `authorization` header is the case that hands over a token of
    /// its own, so its bearer value is unwrapped into the argument.
    fn token(&self, request: &forge_contract::Request) -> Result<Option<String>, String> {
        let mut token = match request.auth {
            Auth::None => None,
            Auth::Bearer => Some(self.var("token")?),
            Auth::Query => {
                return Err("an IPC request carries no query string to hold a token; \
                            the token is an argument, and `auth: bearer` is how a \
                            case asks for it"
                    .into())
            }
        };
        for (name, value) in &request.headers {
            if !name.eq_ignore_ascii_case("authorization") {
                return Err(format!(
                    "an IPC request carries no {name:?} header; a case that needs \
                     one belongs under `inapplicable`"
                ));
            }
            let value = interpolate(value, &self.vars)?;
            let bearer = value.strip_prefix("Bearer ").ok_or_else(|| {
                format!("only a Bearer authorization header maps onto IPC, got {value:?}")
            })?;
            token = Some(bearer.trim().to_string());
        }
        Ok(token)
    }

    fn check(
        &self,
        expect: Option<&Expect>,
        res: &ForgeResponse,
        step: usize,
    ) -> Result<(), String> {
        let Some(expect) = expect else { return Ok(()) };
        let at = |e: String| format!("step {step}: {e}");
        if res.status != expect.status {
            return Err(at(format!(
                "expected status {}, got {} ({})",
                expect.status, res.status, res.body
            )));
        }
        if !expect.headers.is_empty() {
            return Err(at("an IPC response carries no headers; a case that \
                            asserts one belongs under `inapplicable`"
                .into()));
        }
        if expect.text.is_some() {
            return Err(at("an IPC response carries a JSON envelope, never raw \
                            text; a case that asserts text belongs under \
                            `inapplicable`"
                .into()));
        }
        if let Some(want) = &expect.body {
            match_value(want, &res.body, &self.vars).map_err(at)?;
        }
        Ok(())
    }

    fn var(&self, name: &str) -> Result<String, String> {
        self.vars
            .get(name)
            .cloned()
            .ok_or_else(|| format!("no ${{{name}}} in the substitution table"))
    }
}

/// The fixture's users, configured the way a deployment configures them:
/// through the `FORGE_AUTH_USERS` parser. Handing the store a name and a
/// secret directly would step over the parse, which is where an argon2 hash's
/// commas land.
fn auth_config(fixture: &Fixture, vars: &Vars) -> AuthConfig {
    let raw = users_env(&fixture.auth, vars).expect("fixture users");
    let mut config = AuthConfig::new(SECRET);
    config.users = forge_core::auth::parse_users(&raw)
        .expect("the fixture users are not a valid FORGE_AUTH_USERS value");
    // The variable carries no roles and the fixture does, so they go back on
    // by name.
    for user in &fixture.auth.users {
        let name = interpolate(&user.name, vars).expect("user name");
        let stored = config
            .users
            .iter_mut()
            .find(|u| u.name == name)
            .expect("every fixture user survives the parse");
        stored.roles = user.roles.clone();
    }
    config
}

/// The behaviour each fixture action must have, per `contract/README.md`. An
/// action the corpus names and this driver does not know is a failure, not a
/// silent gap.
fn register_action(builder: Builder, name: &str) -> Builder {
    match name {
        "echo" => builder.action("echo", |payload, _ctx| async move { Ok(payload) }),
        "publish" => builder.action("publish", |payload: Value, ctx: ActionCtx| async move {
            let topic = payload
                .get("topic")
                .and_then(Value::as_str)
                .unwrap_or("misc")
                .to_string();
            let data = payload.get("data").cloned().unwrap_or(Value::Null);
            ctx.events.publish(&topic, data);
            Ok(json!({"published": true, "topic": topic}))
        }),
        other => {
            panic!("the corpus fixture wants an action this driver has no behaviour for: {other:?}")
        }
    }
}
